rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,445 @@
+
+For the objects discussed in the previous section, we can embed them as tables in one global SQL database. Below are the schemas of the tables:
+
+```
+Table public_keys {
+  id serial [pk]
+  value text [not null]
+  Indexes {
+    value [name: "public_key"]
+  }
+}
+
+Table blocks {
+  id serial [pk]
+  state_hash string [not null]
+  coinbase int [not null]
+  creator string [not null]
+  staged_ledger_hash string [not null]
+  ledger_hash string [not null]
+  global_slot int [not null] // encompasses (epoch, slot)
+  ledger_proof_nonce int [not null]
+  status int [not null]
+  block_length int [not null]
+  block_time bit(64) [not null]
+  Indexes {
+    state_hash [name: "state_hash"]
+    (block_length, global_slot) [name: "block_compare"]
+    block_time [name: "block_compare"]
+  }
+}
+
+  // Consensus_status is an ADT, but with its small state spaces, it can be coerced into an integer to make it easy to store it in a database. Here are the following states:
+  // - pending : The block is in the transition frontier and its value is its block confirmation number. Value[0, (k-1)]
+  // - confirmed : The block has reached finality by passing the root. Value=-1
+  // - failure: The block got removed from the transition frontier. Value=-2
+  // - unknown: After becoming online and bootstrapping, we may not know what will happen to the block. Value=-3
+
+Enum user_command_type {
+  payment
+  delegation
+}
+
+Table user_commands {
+  id serial [pk]
+  hash string [not null, unique]
+  type user_command_type [not null]
+  nonce bit(64) [not null]
+  sender int [not null, ref: > public_keys.id]
+  receiver int [not null, ref: > public_keys.id]
+  amount bit(64) [not null]
+  fee bit(64) [not null]
+  memo string [not null]
+  first_seen bit(64)
+  Indexes {
+    hash [name: "user_command_hash"]
+    sender [name: "user_command_sender"]
+    receiver [name: "user_command_receiver"]
+    (sender, first_seen) [name: "fast_sender_pagination"]
+    (receiver, first_seen) [name: "fast_receiver_pagination"]
+  }
+}
+
+Table fee_transfers {
+  id serial [pk]
+  hash string [not null, unique]
+  fee bit(64) [not null]
+  receiver int [not null, ref: > public_keys.id]
+  first_seen bit(64)
+  Indexes {
+    hash [name: "fee_transfers_hash"]
+    receiver [name:"fee_transfer_receiver"]
+  }
+}
+
+Table receipt_chain_hashes {
+  id serial [pk]
+  hash string [not null, unique]
+  previous_hash string
+  user_command_id string [not null, ref: > user_commands.id]
+  Indexes {
+    user_command_id
+  }
+}
+
+Table block_to_user_commands {
+  block_id int [not null, ref: > blocks.id]
+  transaction_id int [not null, ref: > user_commands.id]
+  receipt_chain_id int [ref: > receipt_chain_hashes.id]
```

Comment:
This model seems to imply that you will *always* have a block for an associated receipt chain, but I don't think that's actually true. We are planning on implementing the ability to go out and fetch receipt chains for blocks you missed, and we wouldn't want the requirement to be that you have to download the whole block. I think a trinary join table is not the right way to model this. Instead, make one binary join table between blocks and transactions, and another binary join table between transactions and receipt chains. This has an added advantage that queries don't need to be aware of the NULLability of the receipt chain field in the join table, which generally makes writing complex joins painful.

Reply:
Sorry if this is not clear, but the model is trying to state that each receipt chain is associated with a block and a transaction. A transaction can technically appear in multiple blocks, but each of these blocks can produce a different receipt chain hash. This is namely because the previous receipt chain hash would be different for each block. I think this tri-association is necessary because of this reason.


I would prefer having this tri-association but we can also split this table into two tables. One table is a join between blocks and transactions where each association has an id. Then, we can have a join table between the blocks-transactions join table and receipt chain table. The latter seems more complicated and I would prefer not to implement that.